README.md
# Helio Ruby Library

### THIS GEM IS NOT PRODUCTION READY
The first production release will be version 1.0.0. Until then, consider this gem unstable as the Helio API is in flux.

## Installation

You don't need this source code unless you want to modify the gem. If you just
want to use the package, just run:

    gem install helio-ruby

If you want to build the gem from source:

    gem build helio-ruby.gemspec

### Requirements

* Ruby 2.0+.

### Bundler

If you are installing via bundler, you should be sure to use the https rubygems
source in your Gemfile, as any gems fetched over http could potentially be
compromised in transit and alter the code of gems fetched securely over https:

``` ruby
source 'https://rubygems.org'

gem 'rails'
gem 'helio-ruby'
```

## Usage

The library needs to be configured with your account's secret key which is
available in your [Helio Dashboard][api-keys]. Set `Helio.api_token` to its
value:

``` ruby
require "helio"
Helio.api_id = "..."
Helio.api_token = "..."

# list participants
Helio::Participant.list()

# retrieve single participant
Helio::Participant.retrieve(
  "id",
)
```

### Per-request Configuration

It's also possible to set a per-request key and/or account:

``` ruby
require "helio"

Helio::Participant.list(
  {},
  :api_token => "...",
  :api_id => "..."
)

Helio::Participant.retrieve(
  "id",
  :api_token => "...",
  :api_id => "..."
)
```

### Configuring a Client

While a default HTTP client is used by default, it's also possible to have the
library use any client supported by [Faraday][faraday] by initializing a
`Helio::HelioClient` object and giving it a connection:

``` ruby
conn = Faraday.new
client = Helio::HelioClient.new(conn)
participant, resp = client.request do
  Helio::Participant.retrieve(
    "id",
  )
end
puts resp.request_id
```

### Configuring CA Bundles

By default, the library will use its own internal bundle of known CA
certificates, but it's possible to configure your own:

    Helio.ca_bundle_path = "path/to/ca/bundle"

### Configuring Automatic Retries

The library can be configured to automatically retry requests that fail due to
an intermittent network problem:

    Helio.max_network_retries = 2

### Configuring Timeouts

Open and read timeouts are configurable:

```ruby
Helio.open_timeout = 30 // in seconds
Helio.read_timeout = 80
```

Please take care to set conservative read timeouts. Some API requests can take
some time, and a short timeout increases the likelihood of a problem within our
servers.

### Logging

The library can be configured to emit logging that will give you better insight
into what it's doing. The `info` logging level is usually most appropriate for
production use, but `debug` is also available for more verbosity.

There are a few options for enabling it:

1. Set the environment variable `HELIO_LOG` to the value `debug` or `info`:
   ```
   $ export HELIO_LOG=info
   ```

2. Set `Helio.log_level`:
   ``` ruby
   Helio.log_level = Helio::LEVEL_INFO
   ```

### Writing a Plugin

If you're writing a plugin that uses the library, we'd appreciate it if you
identified using `#set_app_info`:

    Helio.set_app_info("MyAwesomePlugin", version: "1.2.34", url: "https://myawesomeplugin.info");

This information is passed along when the library makes calls to the Helio
API.

## Development

Run the linter:

    bundle exec rake rubocop

Update bundled CA certificates from the [Mozilla cURL release][curl]:

    bundle exec rake update_certs

<!--
# vim: set tw=79:
-->


CHANGELOG.md
# Changelog

## 1.0 - 2010-01-05
* 1 major enhancement:
  * Initial release


lib/helio.rb
# Helio Ruby bindings
# API spec at https://helio.zurb.com
require "cgi"
require "faraday"
require "json"
require "logger"
require "openssl"
require "rbconfig"
require "securerandom"
require "set"
require "socket"
require "uri"

# Version
require "helio/version"

# API operations
require "helio/api_operations/create"
require "helio/api_operations/delete"
require "helio/api_operations/list"
require "helio/api_operations/nested_resource"
require "helio/api_operations/request"
require "helio/api_operations/save"

# API resource support classes
require "helio/errors"
require "helio/util"
require "helio/helio_client"
require "helio/helio_object"
require "helio/helio_response"
require "helio/list_object"
require "helio/api_resource"
require "helio/singleton_api_resource"

# Named API resources
require "helio/customer_list"
require "helio/participant"

module Helio
  DEFAULT_CA_BUNDLE_PATH = File.dirname(__FILE__) + "/data/ca-certificates.crt"

  @app_info = nil

  @api_base = "https://helio.zurb.com/api/public"

  @log_level = nil
  @logger = nil

  @max_network_retries = 0
  @max_network_retry_delay = 2
  @initial_network_retry_delay = 0.5

  @ca_bundle_path = DEFAULT_CA_BUNDLE_PATH
  @ca_store = nil
  @verify_ssl_certs = true

  @open_timeout = 30
  @read_timeout = 80

  class << self
    attr_accessor :api_id, :api_token, :api_base, :verify_ssl_certs, :api_version, :client_id,
                  :open_timeout, :read_timeout

    attr_reader :max_network_retry_delay, :initial_network_retry_delay
  end

  # Gets the application for a plugin that's identified some. See
  # #set_app_info.
  def self.app_info
    @app_info
  end

  def self.app_info=(info)
    @app_info = info
  end

  # The location of a file containing a bundle of CA certificates. By default
  # the library will use an included bundle that can successfully validate
  # Helio certificates.
  def self.ca_bundle_path
    @ca_bundle_path
  end

  def self.ca_bundle_path=(path)
    @ca_bundle_path = path

    # empty this field so a new store is initialized
    @ca_store = nil
  end

  # A certificate store initialized from the the bundle in #ca_bundle_path and
  # which is used to validate TLS on every request.
  #
  # This was added to the give the gem "pseudo thread safety" in that it seems
  # when initiating many parallel requests marshaling the certificate store is
  # the most likely point of failure (see issue #382). Any program attempting
  # to leverage this pseudo safety should make a call to this method (i.e.
  # `Helio.ca_store`) in their initialization code because it marshals lazily
  # and is itself not thread safe.
  def self.ca_store
    @ca_store ||= begin
      store = OpenSSL::X509::Store.new
      store.add_file(ca_bundle_path)
      store
    end
  end

  # map to the same values as the standard library's logger
  LEVEL_DEBUG = Logger::DEBUG
  LEVEL_ERROR = Logger::ERROR
  LEVEL_INFO = Logger::INFO

  # When set prompts the library to log some extra information to $stdout and
  # $stderr about what it's doing. For example, it'll produce information about
  # requests, responses, and errors that are received. Valid log levels are
  # `debug` and `info`, with `debug` being a little more verbose in places.
  #
  # Use of this configuration is only useful when `.logger` is _not_ set. When
  # it is, the decision what levels to print is entirely deferred to the logger.
  def self.log_level
    @log_level
  end

  def self.log_level=(val)
    # Backwards compatibility for values that we briefly allowed
    if val == "debug"
      val = LEVEL_DEBUG
    elsif val == "info"
      val = LEVEL_INFO
    end

    if !val.nil? && ![LEVEL_DEBUG, LEVEL_ERROR, LEVEL_INFO].include?(val)
      raise ArgumentError, "log_level should only be set to `nil`, `debug` or `info`"
    end
    @log_level = val
  end

  # Sets a logger to which logging output will be sent. The logger should
  # support the same interface as the `Logger` class that's part of Ruby's
  # standard library (hint, anything in `Rails.logger` will likely be
  # suitable).
  #
  # If `.logger` is set, the value of `.log_level` is ignored. The decision on
  # what levels to print is entirely deferred to the logger.
  def self.logger
    @logger
  end

  def self.logger=(val)
    @logger = val
  end

  def self.max_network_retries
    @max_network_retries
  end

  def self.max_network_retries=(val)
    @max_network_retries = val.to_i
  end

  # Sets some basic information about the running application that's sent along
  # with API requests. Useful for plugin authors to identify their plugin when
  # communicating with Helio.
  #
  # Takes a name and optional version and plugin URL.
  def self.set_app_info(name, version: nil, url: nil)
    @app_info = {
      name: name,
      url: url,
      version: version,
    }
  end

  # DEPRECATED. Use `Util#encode_parameters` instead.
  def self.uri_encode(params)
    Util.encode_parameters(params)
  end
  private_class_method :uri_encode
  class << self
    extend Gem::Deprecate
    deprecate :uri_encode, "Helio::Util#encode_parameters", 2016, 1
  end
end

Helio.log_level = ENV["HELIO_LOG"] unless ENV["HELIO_LOG"].nil?
